### PR TITLE
Fix some type hinting to help with migrating `Distribution`

### PR DESCRIPTION
**Pull request by thomasaarholt**
## Description
This is a small PR. I've added and fixed type hinting to a few lines of code. The idea is to make it easier to migrate the `Distribution` class to a function-based approach as discussed in the related issue.

## Related Issue
- [x] Related to #6083

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7484.org.readthedocs.build/en/7484/

### CODE REVIEW

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
mypy is not happy with the changes? https://github.com/pymc-devs/pymc/actions/runs/10613016885/job/29415940076?pr=7484

**Comment by thomasaarholt** [comment]
**Trying a solution, see below.**

Yes. That's [this commit](https://github.com/pymc-devs/pymc/pull/7484/commits/505effa5dc110d4f1cfc77136f3be59b7126a4dd), where I fixed the incorrect type hint syntax. 

All three reported errors are the same:
```
Incompatible types in assignment 
(expression has type "Callable[[Any, VarArg(Any), DefaultNamedArg(Any, 'size')], Any]", 
base class "Distribution" defined the type as "RandomVariable | SymbolicRandomVariable")
```

and are due to `Mixture`, `Censored` and `Discrete` distributions having `rv_op = <a method>`, while e.g. `Normal` has `rv_op = <a class>`. 

```python
class Censored(Distribution):
    rv_type = CensoredRV
    rv_op = CensoredRV.rv_op # <-- a method

class Normal(Distribution):
    rv_op = NormalRV() # <-- an instance of NormalRV, which is a subclass of RandomVariable
```

**Solution**
Typing the above, I realize that since we are really wanting to type this *by behaviour* rather than *state*. So I've tried typing it as Callable[..., TensorVariable] instead, which is how it always behaves **(I think)**.

**Comment by ricardoV94** [comment]
sounds correct

**Comment by thomasaarholt** [comment]
I am not going to worry about the error in `func_utils.py`. It's in `find_constrained_prior`, which is a) deprecated and b) not typed correctly (first argument should be `type[Distribution]`).

I am currently testing if I can forego the `= None` on the `rv_op` and `rv_type` completely. I don't think these are necessary.

**Comment by ricardoV94** [comment]
> I am not going to worry about the error in `func_utils.py`. It's in `find_constrained_prior`, which is a) deprecated and b) not typed correctly (first argument should be `type[Distribution]`).
> 
> I am currently testing if I can forego the `= None` on the `rv_op` and `rv_type` completely. I don't think these are necessary.
> 

Then can you remove the type hints, we can't merge with a known failing CI or it will be a source of confusion for everyone else

**Comment by thomasaarholt** [comment]
I can remove the type hint for class attribute `rv_op`, but then I either need to leave it `= None` (which the passing tests just established is unnecessary), replace it (with `Any`, probably, and a comment explaining what I've learnt about it), or remove the entire attribute, since the following is not legal Python:

```python
class Foo:
    bar
```

**Comment by thomasaarholt** [comment]
@ricardoV94 mind taking a look?

File: pymc/distributions/distribution.py [hunk 1]
```diff
@@ -122,6 +122,8 @@ def __new__(cls, name, bases, clsdict):
                     )
 
             class_change_dist_size = clsdict.get("change_dist_size")
+            if class_change_dist_size is not None:
+                raise ValueError("HAHAHA")
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Fun debug statement left :)
```suggestion
```

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc/distributions/continuous.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ class BoundedContinuous(Continuous):
     """Base class for bounded continuous distributions"""
 
     # Indices of the arguments that define the lower and upper bounds of the distribution
-    bound_args_indices: list[int] | None = None
+    bound_args_indices: tuple[int, int] | None = None
```
Comment (ricardoV94) [inline pymc/distributions/continuous.py]: Aren't there distributions with only one of the bounds (lower, upper)?

**Review by ricardoV94 — APPROVED** [review]

File: pymc/distributions/continuous.py (hunk 2 above)
Comment (thomasaarholt) [reply pymc/distributions/continuous.py]: You're right, [there are](https://github.com/thomasaarholt/pymc/blob/157c29a60edbe670e245b1208e3e97749669da8d/pymc/distributions/continuous.py#L2084)! I was wondering why mypy didn't flag it, but of course it was originally an error as `list[int]` doesn't allow `(3, None)`! Fixing this

File: pymc/distributions/distribution.py (hunk 1 above)
Comment (thomasaarholt) [reply pymc/distributions/distribution.py]: Hehe! That's what I get for not adding my changes line by line for this particular commit!

**Comment by thomasaarholt** [comment]
The failing tests stem from [this commit](https://github.com/pymc-devs/pymc/commit/465d8ac1675f65f28051423bd7fd84dc5cd2a715), not this PR.

**Comment by ricardoV94** [comment]
@thomasaarholt should be fixed by #7529

**Review by twiecki — APPROVED** [review]

**Comment by ricardoV94** [comment]
Thanks @thomasaarholt

**Pull request merged**